thewalrus/internal_modes/distinguishable_pnr_prob.py

```diff
@@ -0,0 +1,254 @@
+def guan_code(n):
+    """
+    generator for a Guan code sequence
+
+    Args:
+        n (array[int]): maximum value for each digit in the Guan code
+    Yields:
+        j_new (int): index of the changed digit
+        j_pm (int): +1 / -1 depending on whether the index increases or decreases
+    """
+    n = n[::-1]  # maximum for each digit
+    K = len(n)
+    g = np.zeros(K + 1, dtype=np.uint8)  # the guan code
+    u = np.ones(K + 1, dtype=np.int8)
+    while g[K] == 0:
+        # enumerate next gray code
+        i = 0
+        k = g[0] + u[0]
+        while k > n[i] or k < 0:
+            u[i] = -u[i]
+            i += 1
+            k = g[i] + u[i]
+            if i >= K:
+                break
+        g[i] = k
+
+        if i < K:
+            yield K - i - 1, u[i], g[-2::-1]
+
+
+# @numba.jit(nopython=True, cache=True)
+def _dist_prob_gray(pattern, covs, M, hbar=2):
+    r"""
+    probability for distinguishable squeezing GBS.
+
+    Uses some Gray/Guan code tricks to calculate eigenvalues at each step in O(N) time for each squeezer
+
+    algorithm created by Jake Bulmer and Nicolas Quesada (currently unpublished)
+
+    Args:
+        pattern (array[int]): photon number outcome
+        covs (array[:,:,:]): array of covariance matrices, one for each squeezer
+        M (int): number of modes
+        hbar (float): the value of hbar (default 2)
+    """
+    # pylint: disable=too-many-statements
+
+    N = 2 * pattern.sum()
+
+    D = len(covs)
+
+    vac_prob = np.float64(1.0)
+
+    AXs = []
+    abs2_AXs = []
+    assert len(pattern) == M
+    for cov in covs:
+        assert cov.shape[0] == 2 * M
+        Q = nb_Qmat(cov, hbar=hbar)
+        O = np.identity(2 * M) - np.linalg.inv(Q)
+        A = np.empty_like(O)
+        A[:M, :] = O[M:, :].conj()
+        A[M:, :] = O[:M, :].conj()
+        AX = np.empty_like(O)
+        AX[:, :M] = A[:, M:]
+        AX[:, M:] = A[:, :M]
+        # As.append(A)
+        AXs.append(AX)
+
+        abs2_AX = np.real(AX * AX.conj()).astype(np.float64)
+        abs2_AXs.append(abs2_AX)
+
+        vac_prob /= np.sqrt(np.linalg.det(Q).real)
+
+    fac_prod = np.float64(1.0)
+    for p in pattern:
+        fac_prod = fac_prod * np.float64(fact[p])
+
+    traces = np.zeros(D, dtype=np.float64)
+    abs2_AX_sums = np.zeros(D, dtype=np.float64)
+
+    prev_nonzero_rows = [np.int64(i) for i in range(0)]
+    nonzero_rows = [np.int64(i) for i in range(0)]
+
+    edges_sum = np.int64(0)
+
+    H = np.float64(0.0)
+    for j_new, j_pm, kept_edges in guan_code(pattern):
+        edges_sum += j_pm
+
+        if j_pm == 1:
+            nonzero_rows.append(j_new)
+            nonzero_rows.append(j_new + M)
+        else:
+            nonzero_rows.remove(j_new)
+            nonzero_rows.remove(j_new + M)
+
+        # this could be done a tiny bit faster using gray code tricks
+        binom_prod = 1.0
+        for i in range(M):
+            binom_prod *= nb_binom(pattern[i], kept_edges[i])
```

Comment:
```suggestion
            binom_prod *= nb_binom(np.int64(pattern[i]), np.int64(kept_edges[i]))
```
This fixes a numba problem for me (I think the commented out line 86). Alternatively the data type can be changed inside `nb_binom` which already exists or change `if k in (0, n)` to `if k==0 or k==n` in `nb_binom`. All options seem to work for me.